lib/hsa/mcwamp_hsa.cpp

```diff
@@ -2675,24 +2698,32 @@ class HSADevice final : public KalmarDevice
         memcpy(kernel_source, source, kernel_size);
         kernel_source[kernel_size] = '\0';
 
-        // Deserialize code object.
-        hsa_code_object_t code_object = {0};
-        status = hsa_code_object_deserialize(kernel_source, kernel_size, NULL, &code_object);
-        STATUS_CHECK(status, __LINE__);
-        assert(0 != code_object.handle);
+        // Set up ELF header reader
+        elfio reader;
+        istringstream kern_stream{string{
+            kernel_source,
+            kernel_source + kernel_size}};
+        reader.load(kern_stream);
+
+        // Get ISA from ELF header
+        std::string triple = "amdgcn-amd-amdhsa--gfx";
+        unsigned MACH = reader.get_flags() & 0x000000ff;
+
+        switch(MACH) {
+            case 0x023 : triple.append("701"); break;
+            case 0x02a : triple.append("803"); break;
+            case 0x02c : triple.append("900"); break;
```

Comment:
@aaronenyeshi Can we avoid hard-coding ISA version here? We need to reduce future efforts required to support another new gfxXYZ.

Reply:
I can't seem to include llvm/include/llvm/BinaryFormat/ELF.h without it conflicting (compile error) with /usr/include/elf.h.

Reply:
@whchung I know there are many places we have target number such as lib/clamp-device, lit tests, Cmakes, clang driver. Maybe I should expose supported gfxXYZ in clang, and use the structure from there?

Reply:
@aaronenyeshi / @kzhuravl compared with including LLVM headers I think it'd be better to do this within clang. Not sure how easy would it be though.

Reply:
Can you switch from /usr/include/elf.h to llvm's ELF.h?

Reply:
@AlexVlx Do you know if we can easily make the switch from /usr/include/elf.h to llvm's ELF.h?

Reply:
it looks like .note section in the elf object contains the gfx string, can we get it from there?

Reply:
I highly do not recommend that. .note entry, which contains the gfx string, is going away in ROCM2.0.

Reply:
@aaronenyeshi where is /usr/include/elf.h needed?

Reply:
@scchan I think in lib/hsa/mcwamp_hsa.cpp, there are many instances of `using namespace ELFIO;` and we use elf.h inside hc2/external/elfio/elfio_header.hpp

Reply:
`+` @t-tye 

Maybe a long-term goal, but: should not we switch from using elfio to llvm infrastructure?

Not sure what to do in short-term... Should you redefine flags?

Reply:
We've decided to mirror the llvm's ELF.h instead of using internal llvm header. It'll be easy to maintain instead of following LLVM's moving header.

Reply:
@aaronenyeshi  I don't see where we include elf.h in elfio_header.hpp?

@kzhuravl it doesn't seem that llvm's elf provides an API for parsing the elf object or iterating through the symbols/sections, which elfio provides

Reply:
@scchan , I do not think I agree, take a look at:
https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Object/ELF.h#L52
https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Object/ELFObjectFile.h#L201